### PR TITLE
Setup structs

### PR DESCRIPTION
**Pull request by johanni2000**
Implementation of structs Config, Transformer Weights, Runstate and Transformer.

Tests for Config, Runstate and Transformer Weights. 

Transformer struct has nothing to tests so there are no tests for that.

### CODE REVIEW

File: src/config.jl [hunk 1]
```diff
@@ -0,0 +1,27 @@
+"""
+Configuration of intitial parameters
```
Comment (kleincode) [inline src/config.jl:2]: typo: initial

File: src/transformer.jl [hunk 2]
```diff
@@ -0,0 +1,123 @@
+"""
+Initial parameters for the transformer weights
+
+llama.c correspondence: TrasformerWeights (l. 29)
```
Comment (kleincode) [inline src/transformer.jl:4]: "llama2.c" not "llama.c" (I did the same mistake) + typo: TransformerWeights

File: src/transformer.jl [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    rms_final_weight::Vector{Float32} # (dim,)
+
+    # optional classifier weights
+    # wcls::Array{Float32} 
+end
+
+"""
+Initial parameters for the run state
+
+llama.c correspondence: RunState (l. 50)
```
Comment (kleincode) [inline src/transformer.jl:35]: "llama2.c"

File: src/transformer.jl [hunk 4]
```diff
@@ -0,0 +1,123 @@
+"""
+Initial parameters for the transformer weights
+
+llama.c correspondence: TrasformerWeights (l. 29)
+"""
+struct TransformerWeights    # using @kwdef for clearer parameter initialization in the outer constructor
```
Comment (kleincode) [inline src/transformer.jl:6]: Remove comment as we are not using @kwdef anymore

File: src/transformer.jl [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    # optional classifier weights
+    # wcls::Array{Float32} 
+end
+
+"""
+Initial parameters for the run state
+
+llama.c correspondence: RunState (l. 50)
+"""
+struct RunState  # using @kwdef for clearer parameter initialization in the outer constructor
```
Comment (kleincode) [inline src/transformer.jl:37]: Remove comment as we are not using @kwdef anymore

File: src/transformer.jl [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+    # kv cache
+    key_cache::Array{Float32,3}  # (layer, seq_len, dim)
+    value_cache::Array{Float32,3}  # (layer, seq_len, dim)
+end
+
+"""
+Initial parameters for the transformer
+
+llama.c correspondence: Transformer (l. 67)
```
Comment (kleincode) [inline src/transformer.jl:58]: llama2.c

File: src/transformer.jl [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    fd::Int32 # file descriptor for memory mapping
+    data::Array{Float32} # memory mapped data pointer
+    file_size::Int32 # size of checkpoint file in bytes
+end
+
+"""
+Initialization of RunState based on Config
+- initialized as undefined arrays
+
+llama.c correspondence: malloc_run_state (l. 77)
```
Comment (kleincode) [inline src/transformer.jl:75]: llama2.c

File: src/transformer.jl [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        Array{Float32}(undef, config.n_layers, config.seq_len, kv_dim),
+        Array{Float32}(undef, config.n_layers, config.seq_len, kv_dim),
+    )
+end
+
+"""
+Initialize transformer weights based on Config
+- intitialized as undefined arrays
+
+llama.c correspondence: memory_map_weights (l. 111)
```
Comment (kleincode) [inline src/transformer.jl:103]: llama2.c

File: src/transformer.jl [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        Array{Float32}(undef, config.n_heads, config.seq_len),  # rms_att_weight
+        Array{Float32}(undef, config.vocab_size),   # logits
+        Array{Float32}(undef, config.n_layers, config.seq_len, kv_dim),
+        Array{Float32}(undef, config.n_layers, config.seq_len, kv_dim),
+    )
+end
+
+"""
+Initialize transformer weights based on Config
+- intitialized as undefined arrays
```
Comment (kleincode) [inline src/transformer.jl:101]: typo: initialized

File: test/test_config.jl [hunk 10]
```diff
@@ -0,0 +1,124 @@
+using Llama2
+using Test
+
+@testset "Config" begin
+    @testset "Initialize configuration" begin
+        @testset "Good case" begin
+            # inititalization should not throw an error
```
Comment (kleincode) [inline test/test_config.jl:7]: typo: initialization (multiple occurrences in this file)

File: test/test_transformer.jl [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        n_layers::Int32 = 3
+        n_heads::Int32 = 4
+        n_kv_heads::Int32 = 6
+        vocab_size::Int32 = 30
+        seq_len::Int32 = 2
+
+        config = Config(dim, hidden_dim, n_layers, n_heads, n_kv_heads, vocab_size, seq_len)
+        state = RunState(config)
+
+        kv_dim::Int32 = (dim * n_kv_heads) ÷ n_heads
```
Comment (kleincode) [inline test/test_transformer.jl:50]: Just a comment: I think values should always be chosen such that dim*n_kv_heads is a multiple of n_heads. We will probably find this out later.

**Review by kleincode — CHANGES_REQUESTED** [review]
Code and tests look good, please fix typos

**Review by kleincode — APPROVED** [review]
